### ISSUE TITLE
Score completion thoughts

### ISSUE THREAD

**Issue by sgoshorn** [CCALI/CALI-Author-Viewer-5#68]
The Bitovi score redesign assumes there is a % completion. % completion is separate from % score. Currently % completion is not known due to branching. Any lessons with branching will present problems but showing students how far they are in a lesson is a GREAT IDEA. But just showing the total number of questions in a lesson can be misleading since they may wonder how they reached the end with only 35 out of 40 questions answered. 

Instead, as a first attempt at completion, what if we don’t show numbers at all, just a visual progress element (maybe %). But this number only counts pages in the Next flow. These would be the pages you see in the outline in CALI Author mode which doesn't normally count the branches. There might be 30 questions in the flow with 10 extras in branches.  Students would just sees their % progress from 0 to 100% (internally, 0/30 to 30/30). When they hit a branch the progress bar doesn’t seem to move but once they are back on an outline's page it advances again. When they reach the final page of the lesson they will hit 100%. (Pedagogy issue for @dquentel : if they can tell they are in a branch will this affect how they answer? Would they be discouraged and go Back and try another path?)

Getting the completion count from outline Nexts is doable since this algorithm is already used in CA. (The Jump link problem will be ignored for now.) If a lesson has several parts and students start on the last part then their % completion will not include those first parts. They need to visit every page in the outline Next page flow to count towards completion. Lessons without branching will be 100% accurate. Heavy branching ones may look odd (progress bar advancing only in spurts as they spend most time in branches.) And there are probably outliers with branching and Nexts in conflict which may need author intervention to correct.

This would mean a student could get a score of 100% for answering 2 out of 2 questions correctly but have a completion of 7% (2/30 questions.) They could also jump around and answer every question to answer 40/30 questions but those branch questions don't count towards completion so I think it would stay at 100% rather than 133%.

**Comment by dquentel** [comment]
Pedagogy - to address Sam's question above - > if they can tell they are in a branch will this affect how they answer? Would they be discouraged and go Back and try another path?> 
- they can tell they're in a branch - I don't think this will affect how students answer the question. They're always going to try to get it right. Or, skip it. We can disable next page to eliminate page skipping, if we want. 
- discouraged - perhaps. For more than a decade now, I've had conversations with Fellows about the negative impact on students of jumping them backwards in a lesson if they get a question wrong.  Instead, authors will use a popup to review the material. 
- go back and try another path - I'm not seeing this as a negative. I don't know the real number but I would maybe a third of branching questions already loop students back to the original question to "try again" after a series of branching questions. Since the lessons are about learning, it seems letting the student go back and try different paths should be allowed. 

I really, really like the idea of the percentage progress bar. However, I am concerned with it appearing to "freeze" on a branching series of questions. By showing progress we are providing a level of feedback to the student. And, if that feedback seems to stop, they'll wonder why. Can the math account for this? Maybe following next pages is 20/20 and following all branches is 25/25. The progress bar is based on the 25/25 and then jumps ahead to 100% at the final question. And, even as I type this I see issues with it jumping for lessons with lots of branching. But freezing the progress bar, in my opinion, could cause the student more concern. 

As an example, I know I often take surveys where the opening question is something to filter me into or out of the survey and the progress bar inches forward. But then when the survey rejects me it jumps to the end.

**Comment by sgoshorn** [comment]
I agree that no apparent progress would be concerning. I think a jump in progress would be OK. If we go with the filling circle rather than horizontal bar it would be even less noticeable. Also, maybe we make it explicit if a lesson is linear or branching - an opportunity to promote our non-linear teaching method with a snazzy demo popup? Maybe not.
There seems to be no reliable way to handle this but showing a progress bar seems better than not.  My next thought is to go ahead and show the progress but based on all questions. Linear lessons would work fine. Those with some branching will still advance for every question. Then the issue of being incomplete is resolved with a simple hack: Upon arriving at the last question of the Conclusion section, set progress to 100% no matter what. Only students who start a new lesson on the last section first would notice anything odd. This method would not require the extra step of mapping each page to a section which is then even more complicated by jump links. A linear lesson would show perfect behavior. A heavy branching lesson might show 75% (30/40 questions) right before the last page then suddenly jump to 100% but I think that's OK.
